packages/next-swc/crates/napi/src/util.rs

```diff
@@ -89,3 +93,45 @@ where
     serde_json::from_str(s)
         .with_context(|| format!("failed to deserialize as {}\nJSON: {}", type_name::<T>(), s))
 }
+
+/// Initialize tracing subscriber to emit traces. This configures subscribers
+/// for Trace Event Format (https://docs.google.com/document/d/1CvAClvFfyA5R-PhYUmn5OOQtYMH4h6I0nSsKchNAySU/preview).
+#[js_function(1)]
+pub fn init_custom_trace_subscriber(cx: CallContext) -> napi::Result<JsUndefined> {
+    let optional_trace_out_file_path = cx.get::<JsUnknown>(0)?;
+    let trace_out_file_path = match optional_trace_out_file_path.get_type()? {
+        napi::ValueType::String => Some(PathBuf::from(
+            JsString::try_from(optional_trace_out_file_path)?
+                .into_utf8()?
+                .as_str()?
+                .to_owned(),
+        )),
+        _ => None,
+    };
+
+    CUSTOM_TRACE_SUBSCRIBER.get_or_init(|| {
+        let mut layer = ChromeLayerBuilder::new().include_args(true);
+        if let Some(trace_out_file) = trace_out_file_path {
+            layer = layer.file(trace_out_file);
+        }
+
+        let (chrome_layer, guard) = layer.build();
+        tracing_subscriber::registry()
+            .with(chrome_layer.with_filter(filter::filter_fn(|metadata| {
+                !metadata.target().contains("cranelift") && !metadata.name().contains("log ")
+            })))
+            .try_init()
+            .expect("Failed to register tracing subscriber");
+
+        cx.env
+            .add_env_cleanup_hook(guard, |flush_guard| {
```

Comment:
Q for @Brooooooklyn : unlike `@swc/core`, this cleanup hook doesn't seem to be called. In result, generated traces are not complete (clipped in the end). 

I even tried to throw `panic!` in this hook but it is still not being called. Is there any missing step to configure this work correctly?

Reply:
It's weird... Is it that the process has not been executed when it exits? Is it possible that the webpack or next.js process is still running?

Reply:
I ran bench/nested which completes with one off build. I do not believe we run webpack or else as long-running for background pnce main process exits.

Reply:
I will dig into this issue, when I have a conclusion will come back to sync to you

Reply:
Thanks, appreciate for the help.

Reply:
I've tried this:

```js
const {
  initCustomTraceSubscriber,
} = require('./native/next-swc.darwin-arm64.node')

console.log(initCustomTraceSubscriber())
```

the `add_env_cleanup_hook` was working as expected.

/cc @sokra any idea here?

Reply:
Huh you're right, individual import calls cleanup_hook but if swc is being loaded in webpack (or next?) context hook is not being called. It is the same process, not sure what can affect this.

Reply:
@Brooooooklyn Ok, was able to narrow down finally and I'm not sure if this is expected behavor or not.

https://github.com/vercel/next.js/blob/4db8c49cc31e4fc182391fae6903fb5ef4e8c66e/packages/next/bin/next.ts#L134= is root cause, not from webpack or anything else.

- Working
```
initCustomTraceSubscriber();
transform();

// cleanup called correctly
```

- Not working
```
x.initCustomTraceSubscriber();
x.transform('console.log');

process.exit(0)
// cleanup not being called
```

I was under assumption cleanup should be called regardless of manually exiting process? 🤔

Reply:
Ok, tried bunch and so far I couldn't find any reliable way to call cleanup_hook if parent process explicitly exits, especially doc implies it is some sort of expected behavior

> Calling process.exit() will force the process to exit as quickly as possible even if there are still asynchronous operations pending that have not yet completed fully, including I/O operations to process.stdout and process.stderr.

it doesn't seem to easily achieve.

Instead, switched direction to explicitly closes out subscriber when process exits. So far it looks like working as expected.

Reply:
> I was under assumption cleanup should be called regardless of manually exiting process? 🤔

Not sure if this is expected behavior for Node.js, I'll report it to nodejs

Reply:
https://github.com/nodejs/node/discussions/42609